Add Layout/IndentationConsistency to Rubocop config

herwinw opened the pull request.

herwinw commented:
As suggested in https://github.com/ruby/spec/pull/1174#discussion_r1660135696

herwinw commented on core/basicobject/instance_exec_spec.rb:
```diff
@@ -84,17 +84,17 @@ def foo
-    1.instance_exec { @@__tmp_instance_exec_spec }.should == 2
-    Integer.__send__(:remove_class_variable, :@@__tmp_instance_exec_spec)
+  quarantine! do # Not clean, leaves cvars lying around to break other specs
+    it "scopes class var accesses in the caller when called on an Integer" do
+      # Integer can take instance vars
+      Integer.class_eval "@@__tmp_instance_exec_spec = 1"
+      (defined? @@__tmp_instance_exec_spec).should == nil
+
+      @@__tmp_instance_exec_spec = 2
+      1.instance_exec { @@__tmp_instance_exec_spec }.should == 2
+      Integer.__send__(:remove_class_variable, :@@__tmp_instance_exec_spec)
+    end
   end
-end
```
I'm not totally sure about these `quarantine!` blocks, they're kind of special, do we want to keep the special indentation for those?

eregon replied:
Right for `quarantine!` we sometimes avoided to reindent to minimize changes and noise in diffs/git log -p.
I think it would be nice to avoid indenting more under `quarantine!`, but does RuboCop support that?

eregon replied:
It looks like there is no way to configure that from a quick look at https://docs.rubocop.org/rubocop/cops_layout.html#configurable-attributes-31
It's fine to reindent, we don't quarantine/unquarantine often.

herwinw replied:
There are two ways to work around this:
* Add an exclude for the specific cop for this file, which means we ignore this specific cop in the whole file (not just this quarantine line)
* Add a `rubocop:disable Layout/IndentationConsistency` line in the source to make Rubocop ignore the specific line or block. This means we're mixing up Rubocop with the code

I think both of these options are pretty ugly, so I would be okay with not added any exceptions here.

eregon replied:
Yeah let's just reindent as you did.
In general I would prefer per-file exclude instead of rubocop comments, but even better if we don't have to exclude at all.

herwinw commented on core/integer/pow_spec.rb:
```diff
@@ -19,13 +19,13 @@
-       2.pow(5,  12).should ==  8
-       2.pow(5, -12).should == -4
-      -2.pow(5,  12).should ==  4
+      2.pow(5, 12).should == 8
+      2.pow(5, -12).should == -4
+      -2.pow(5, 12).should == 4
       -2.pow(5, -12).should == -8
```
Rubocop trips here because the ` 2` and `-2` are aligned. I just brutally ignored that, do we want a special case for this? I think there is a Rubocop setting for this, but I have to look up which one it is.

eregon replied:
It's good as-is, I like simple alignment, much easier to maintain

herwinw commented on core/proc/curry_spec.rb:
```diff
@@ -159,10 +159,12 @@
-      raise_error(ArgumentError)
-    -> { @proc_add.curry(1)[1,2].curry(3)[3,4,5,6].should == 6 }.should_not
-      raise_error(ArgumentError)
+    -> {
+      @proc_add.curry(3)[1,2,3,4].should == 6
+    }.should_not raise_error(ArgumentError)
+    -> {
+      @proc_add.curry(1)[1,2].curry(3)[3,4,5,6].should == 6
+    }.should_not raise_error(ArgumentError)
```
Here it felt better to just write them in the style of most of these blocks, that resolves the rubocop issue too.

herwinw replied:
It looks like this is an actual change in behaviour (I guess the old version parsed it as two statements), which would indicate this test is actually broken.

herwinw replied:
I'm not sure why we even would want a `should_not raise_error` around a `should` statement, this feels superfluous here.

eregon replied:
Yeah `should_not raise_error` is an anti-pattern, let's remove these.

eregon reviewed (approved):
Thank you!

herwinw commented:
I've added a few additional commits. This includes a `Gemfile` with the correct Rubocop version (the `Gemfile.lock` is ignored in git, so I left that one out) and updated Github Actions to use the gemfile. This means it's easier to run Rubocop on your home setup without having to worry about installing the correct version of Rubocop.

eregon commented:
Mmh, one trouble with the Gemfile is if someone looks at this project they might think they should run `bundle exec mspec` or so to run specs, because there is a Gemfile at the root.
But on purpose ruby/spec avoids Bundler because subprocesses are a lot slower under Bunder and to keep things simple.

What I do locally is basically the same the CI does:
https://github.com/ruby/spec/blob/8aae62fbdc0e646ae775576d847beb60063ca253/.github/workflows/ci.yml#L75C12-L76
Except for the second command I use:
```
$ rubocop _1.28.2_ --parallel
```
(we could of course use that in CI too to be more explicit)

So I think either using that or adding a small script which just runs `rubocop _1.28.2_ *args` is better than the Gemfile.

eregon commented:
Also since RuboCop 1.0 I think using a newer version of RuboCop has almost no observable behavior difference, so that should work fine too.

eregon commented:
@herwinw Could you address https://github.com/ruby/spec/pull/1180#issuecomment-2408628698 ?
I think having a Gemfile at the root is not good because of those reasons so I would like to find another solution.

herwinw commented:
> @herwinw Could you address [#1180 (comment)](https://github.com/ruby/spec/pull/1180#issuecomment-2408628698) ? I think having a Gemfile at the root is not good because of those reasons so I would like to find another solution.

My apologies, I completely overlooked that one. I will have a look at it tomorrow or the day after tomorrow.
